dyndns/domain_update.py

```diff
@@ -27,37 +28,98 @@ def main(domain, settings='settings.txt', ignore_previous_ip=False):
+                if not answers[record_type]:
+                    return "No {} record found for domain {}".format(record_type, domain)
+                ip[proto] = answers[record_type][0]
+                print("IP {} found in {} record".format(ip[proto], record_type))
+
+            except Exception:
+                print("No {} record found for domain {}".format(record_type, domain))
+
+        config[domain]['last_dns_check'] = int(time.time())
+
+    public_ip = {}
+    for proto in protocols:
+
+        # get public ip
+        response = requests.get(protocols[proto]['api'], params={'format': 'json'})
         try:
-            answers = dns.resolver.query(domain, 'A')
-            if not answers:
-                return "No A record found for domain {}".format(domain)
-            ip = answers[0]
-            print("IP {} found in A record".format(ip))
-            config[domain]['last_dns_check'] = int(time.time())
-        except Exception:
-            print("No A record found for domain {}".format(domain))
-
-    # get public ip
-    response = requests.get("https://api.ipify.org", params={'format': 'json'})
-    if response.status_code != 200:
-        return "Could not discover public IP."
-    public_ip = response.json().get('ip', None)
-    if not public_ip:
-        return "Could not discover public IP."
-
-    print("New IP: {}".format(public_ip))
-    if not ignore_previous_ip and ip and str(ip) == str(public_ip):
-        config[domain]['ip'] = public_ip
-        # update settings
-        with open("settings.txt", "w") as settings_file:
-            json.dump(config, settings_file, sort_keys=True, indent=1)
-        return "A record up to date."
+            response_ip = response.json().get('ip', None)
+            # validate http response code
+            if response.status_code != 200:
+                raise ValueError("Could not discover public {} address.".format(proto))
+
+            # validate the correct returned IP Version
+            if ipaddress.ip_address(response_ip.decode('unicode-escape')).version != protocols[proto]['version']:
+                raise ValueError('The recieved IP address {} doesn\'t match the Protocol {}'
+                                 .format(response_ip, proto))
+
+            # validation ok
+            public_ip[proto] = response_ip
+            print("New {} address: {}".format(proto, public_ip[proto]))
+
+        except ValueError as error:
+            print(error)
+            # validation failed! do we have a valid ip elsewhere?
+            # if whe got an ip from the dns query, use it
+            if proto in ip and ip[proto]:
+                public_ip[proto] = ip[proto]
+                print("Fallback to recieved {} address from dns query: {}"
+                      .format(proto, str(ip[proto])))
+            # if whe still have an ip in the config, use it
+            elif ('ip' in config[domain]
+                  and proto in config[domain]['ip']
```

Comment:
ditto, not sure of value/reason and side effects for this fallback.
wouldn't be better just to abort if public_ip not received?

Reply:
Same as above, your decision :)

Reply:
Commented above. I also tried another approach which should work, to selectively update only one group, if IP of the other group is not available at the moment.
I have it on a branch as of now, because there seems to be a bug in the IONOS implementation (such update fails). I reported to the dev team and let's see what they say. We may then keep of remove the fallback compltely.